### ISSUE TITLE
19-Shape-Guided Diffusion with Inside-Outside Attention

### ISSUE THREAD

**Issue by mjmaher987** [sharif-ml-lab/visual_compositional_generation_survey#19]
## Task: Read the Paper and Do the following 4 steps
1. Highlight the paper systematically (**green** for important notes - **blue** for very important notes that worth mentioning in the final survey - **red** for the sentences that we don’t understand or need further attention in the future) [In Zotero].
2. Write the Wikipedia summary of the paper in the GitHub issue (_comment in this issue_).
3. Extract the information in the [excel file](https://docs.google.com/spreadsheets/d/15_sh6TqtUvZIJo1-RZxQtsG0Nvg0vPG_IFKCWiGMjdY/edit?usp=sharing).
4. Add the related sentences to the [overleaf file](https://www.overleaf.com/8167246262ccxtrzztjvpz). Note that when you add the sentence, don’t forget to write the reference in the format of ”[ #reference number that is mentioned in the Google Doc - Name of the Author (who added this to the overleaf) ]” at the end of the sentence.

**Comment by AlieNiT** [comment]
# Title: Shape-Guided Diffusion with Inside-Outside Attention

<img width="1136" alt="Screenshot 2023-11-23 at 6 29 26 AM" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/80975118/4a968ae4-d3a6-4e66-8c24-ebb552d5b6f7">

## Abstract
- In object manipulation tasks, existing text-to-image diffusion models overlook object's shape, resulting in inaccuracies such as incorrect scaling, cutoffs, or replacement with background content during object manipulation.
- The paper introduces a training-free approach called Shape-Guided Diffusion, enabling modification of pre-trained diffusion models to consider shape input provided by users or inferred from text.
- They implemented a novel mechanism named as Inside-Outside Attentio during inversion and generation processes within cross- and self-attention maps, enforcing the shape constraint.
- This mechanism distinguishes object (inside) from background (outside) spatial regions and associates text prompt edits with the correct region, enhancing object replacement based on user prompts.
- They demonstrate the efficacy of their method on the shape-guided editing task, where the model must replace an object according to a text prompt and object.
- They've also created a new ShapePrompts benchmark from MS-COCO dataset and achieved state-of-the-art (SOTA) results in shape fidelity while maintaining text alignment and image realism, as validated by both automatic metrics and annotator ratings.
- website: https://shape-guided-diffusion.github.io
## Introduction
- Recent advancements in large-scale diffusion models have notably enhanced the realism and fidelity of text-driven image synthesis but face limitations in expressing complex specifications solely through language (prompt).
- Previous image editing methods often fail to maintain original object shapes, resulting in deficient affordances (producing a dog with missing limbs) or interactions (inconsistent light reflections).
- The proposed method addresses spurious attentions within diffusion models by employing an innovative Inside-Outside Attention mechanism, distinguishing object and background pixels during inversion and generation, enhancing the preservation of object shapes.
Contributions highlighted:
  1.  Identification of a limitation in previous image editing techniques regarding preservation of original object shapes and empirical insights into the issue's causes.
  2.  Introduction of a training-free mechanism enforcing shape constraints on attention maps during inference, a novel approach compared to existing mask-based adaptations.
  3.  Achievement of state-of-the-art (SOTA) results in shape fidelity on the MS-COCO ShapePrompts benchmark, surpassing competitive baselines and receiving higher annotator ratings for editing capabilities such as object, background, and simultaneous inside-outside edits.

## Related Work

### Diffusion Models

  - Remarkable success achieved in image synthesis via diffusion models, establishing a Markov chain diffusion process that adds noise and constructs desired data samples.
  - Variants like Denoising Diffusion Probabilistic Models (DDPM), Denoising Diffusion Implicit Models (DDIM), and score-based models are prominent.
  - Previous exploration of classifier guidance and classifier-free guidance for conditional image synthesis.
  - Recent focus on text-guided image synthesis within diffusion models, motivating adaptation towards text-guided local editing using text prompts and object masks.
  - Mention of the limitations of existing methods like Blended Diffusion, emphasizing its insufficiency in preserving object shape, prompting the need for improvements demonstrated in the presented method.

### Global and Local Image Editing

  - Overview of methods integrating generative models for image editing, particularly text-guided global editing with adaptations like StyleCLIP and DiffusionCLIP.
  - Description of Blended Diffusion's technique for local editing constrained to a mask, highlighting its limitations in preserving object shape, which is addressed by the proposed method.

### Structure Preserving Image Editing
  - Discussion on works dedicated to structure-preserving image editing, emphasizing the preservation of object position or pose during edits.
  - Various approaches including copying random seeds, finetuning model weights, copying features and self-attention maps, conditioning on noised source images, and methods like PrompttoPrompt (P2P).
  - Comparison between prior works and the proposed method regarding attention maps' utilization and sensitivity to shape guidance.
  - Exploration of issues faced by other methods in preserving structure in real images and the proposed method's effectiveness in mitigating these issues through attention map constraints.

### Image Inpainting
  - Overview of image inpainting techniques utilizing various approaches like dilated convolution, partial convolution, gated convolution, contextual attention, and co-modulation.
  - Mention of diffusion-based models like GLIDE and Stable Diffusion adapted for text-conditional inpainting, highlighting their training on free-form masks without semantic meaning.
  - Description of training-based methods using object masks for inpainting, including Make-a-Scene, Shape-guided Object Inpainting, and Imagen Editor.
  - Distinction of the presented method's training-free approach in text-to-image diffusion models at inference time, enabling flexibility for diverse tasks beyond object editing.

## Shape-Fuided Diffusion (Proposed Method)

- Introduction of Shape-Guided Diffusion as a training-free method designed to enable a pre-trained text-to-image diffusion model to adhere to shape guidance.
- Aim: Local editing of image $X_{src}$ using text prompts $P_{src}$ and $P_{edit}$, alongside an optional object mask $m$ (inferred from $P_{src}$ if absent), to produce edited image $X_{edit}$ faithful to both $P_{edit}$ and $m$.
- Implementation of Inside-Outside Attention to explicitly regulate cross- and self-attention maps during both the inversion and generation processes, ensuring shape constraints.
- Leveraging Stable Diffusion (SD), a Latent Diffusion Model (LDM) functioning in a low-resolution latent space derived from the image space, facilitating the application of Inside-Outside Attention via downsampled object masks within this space.
- Clarification that "pixel," "image," or "noise" references in this context relate to concepts within the LDM latent space, emphasizing the perceptual equivalence of this space to the downsampled version of the image space. The mask is also downsampled to match latent dimensions.
### Inside-Outside Attention
#### Algorithm
<img width="547" alt="Screenshot 2023-11-23 at 3 17 59 PM" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/80975118/f22a5859-5dd9-4243-b298-97a26e5840a0">

#### Visual Interpretation
<img width="547" alt="Screenshot 2023-11-23 at 3 10 50 PM" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/80975118/456cdf5c-809b-465f-aca6-567e16a36bc5">

### Inside-Outside Inversion
- Utilization of DDIM inversion to transform source images into inverted noise for editing real images, observed to cause a potential tradeoff between shape-text faithfulness in conjunction with text-to-image diffusion models.
- Concerns arise regarding the impact of nonzero levels of classifier-free guidance, potentially leading to the loss of information about the source object, depicted in the bottom row of below figure.
- Explanation of how the method maintains consistency: if a token remains constant (from $P_{src}$ to $P_{edit}$), its associated region remains unchanged (e.g., "dog" associated with "background" in reconstruction); if the token changes, its associated region changes (e.g., "dog" in the editing setting), without affecting other regions.
- Comparison between outcomes: Without the proposed method (bottom), inverted noise retains some information but results in changes unrelated to the original image (e.g., chair's checkerboard pattern transformed into flowers or polka dots). With the method (top), the edited image retains the complete chair.
- Depiction of the cross-attention map for "dog," showcasing weak inclusion of the background without the proposed method and localized attention specifically to the dog with the method's implementation.

### Visual Interpretation
<img width="935" alt="Screenshot 2023-11-24 at 2 37 36 AM" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/80975118/2cf708fa-fd44-4462-b024-62de0fb012ce">

### Method Summary

#### Algorithm
<img width="575" alt="Screenshot 2023-11-24 at 2 47 10 AM" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/80975118/30d46cbd-0434-4e1d-a8f0-0a618da9e4f7">

## MS-COCO ShapePrompts (Dataset and Metrics)

- Evaluation conducted on MS-COCO images involving a test set derived from MS-COCO val 2017 containing 1,149 object masks covering 10 categories such as animal, vehicle, food, and sports classes. Additionally, a validation set with 1,000 object masks from MS-COCO train 2017 was created in a similar fashion.
- Designed prompts for various categories involving clothing, color manipulation, material change, and specifying rare subcategories to manipulate and augment images.
- Metrics employed for evaluation:
  - Mean Intersection over Union (mIoU): Measures the proportion of pixels within the masked region correctly synthesized as the desired object class, determined by a segmentation model trained on COCO-Stuff.
  - Keypoint-weighted mIoU (KW-mIoU): Specifically for fine-grained animal classes, weights each sample’s mIoU by the percentage of correct keypoints when comparing source vs. edited images, as determined by an animal keypoint detection model.
  - FID scores: Evaluate image realism by measuring the similarity between distributions of real and synthetic images using Inception network features.
  - CLIP scores: Assess image-text alignment by measuring the similarity between the text prompt and synthetic image using features from a large pretrained image-text model.
- Emphasis on mIoU for faithful object synthesis, with additional consideration for animal classes using KW-mIoU due to their fine-grained nature. FID scores gauge image realism, while CLIP scores evaluate image-text alignment.
Further detailed information on these metrics can be found in the supplemental material.

## Experiments

### Annotator evaluation
<img width="1130" alt="Screenshot 2023-11-24 at 3 36 58 AM" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/80975118/78c7274f-639f-4d02-b16a-5409f18f1a7f">

### Automatic Evaluation (and ablation)
<img width="555" alt="Screenshot 2023-11-24 at 3 37 21 AM" src="https://github.com/sharif-ml-lab/visual_compositional_generation_survey/assets/80975118/42c00260-84e5-497d-af89-1e446bfcbbfa">

## Concolusion


- Emphasizing the significance of explicit shape consideration for local edits on real images.
- Identification of limitations in prior methods related to local editing, structure-preserving editing, and inpainting due to their failure to respect shape constraints.
- Introduction of Shape-Guided Diffusion, a training-free approach leveraging Inside-Outside Attention during inversion and generation processes to localize object vs. background edits.
- Evaluation conducted on the proposed MS-COCO ShapePrompts benchmark for shape-guided editing tasks, demonstrating the superior performance of the proposed method compared to baselines in shape faithfulness.
- Highlighting the method's success in maintaining shape faithfulness without compromising text alignment or image realism, whether using precise MSCOCO instance masks or masks inferred from the text.
